// @playzone/youtube-transcript/dist/enhanced-api/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnhancedYouTubeTranscriptApi = void 0;
const axios_1 = __importDefault(require("axios"));
const http_proxy_agent_1 = require("http-proxy-agent");
const https_proxy_agent_1 = require("https-proxy-agent");
const api_1 = require("../api");
const USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36,gzip(gfe)';
/**
 * Enhanced YouTube Transcript API with advanced proxy and Invidious support
 */
class EnhancedYouTubeTranscriptApi {
    constructor(proxyOptions = {}, invidiousOptions = {}) {
        this.invidiousClient = null;
        // Default proxy options
        this.proxyOptions = {
            enabled: false,
            http: '',
            https: '',
            ...proxyOptions,
        };
        // Default Invidious options
        this.invidiousOptions = {
            enabled: false,
            instanceUrls: '',
            timeout: 10000,
            ...invidiousOptions,
        };
        // Initialize base API
        this.baseApi = new api_1.YouTubeTranscriptApi();
        // Initialize HTTP client with enhanced proxy support
        this.initializeHttpClient();
        // Initialize Invidious client if enabled
        if (this.invidiousOptions.enabled) {
            this.initializeInvidiousClient();
        }
    }
    /**
     * Initialize HTTP client with enhanced proxy support
     */
    initializeHttpClient() {
        const config = {
            headers: {
                'Accept-Language': 'en-US',
                'User-Agent': USER_AGENT,
                'Accept-Encoding': 'gzip, deflate, br',
            },
            timeout: 10000,
            maxRedirects: 5,
        };
        // Add proxy configuration if enabled
        if (this.proxyOptions.enabled) {
            config.proxy = false; // Disable built-in proxy resolver
            // Only use proxy agents in Node.js environments
            if (typeof globalThis.window === 'undefined') {
                config.httpAgent = new http_proxy_agent_1.HttpProxyAgent(this.proxyOptions.http || '');
                config.httpsAgent = new https_proxy_agent_1.HttpsProxyAgent(this.proxyOptions.https || this.proxyOptions.http || '');
            }
        }
        else if (typeof globalThis.window === 'undefined') {
            // Use keep-alive agents when not using proxy
            const http = require('http');
            const https = require('https');
            config.httpAgent = new http.Agent({ keepAlive: true });
            config.httpsAgent = new https.Agent({ keepAlive: true });
        }
        this.httpClient = axios_1.default.create(config);
    }
    /**
     * Initialize Invidious client
     */
    initializeInvidiousClient() {
        const instanceUrls = Array.isArray(this.invidiousOptions.instanceUrls)
            ? this.invidiousOptions.instanceUrls
            : [this.invidiousOptions.instanceUrls];
        if (instanceUrls.length === 0 || (instanceUrls.length === 1 && !instanceUrls[0])) {
            throw new Error('At least one Invidious instance URL must be provided when Invidious is enabled');
        }
        const primaryInstanceUrl = instanceUrls[0];
        const config = {
            baseURL: primaryInstanceUrl,
            timeout: this.invidiousOptions.timeout || 10000,
            headers: {
                Accept: 'application/json',
                'User-Agent': USER_AGENT,
            },
        };
        // Add proxy configuration if enabled
        if (this.proxyOptions.enabled && typeof globalThis.window === 'undefined') {
            config.proxy = false;
            config.httpAgent = new http_proxy_agent_1.HttpProxyAgent(this.proxyOptions.http || '');
            config.httpsAgent = new https_proxy_agent_1.HttpsProxyAgent(this.proxyOptions.https || this.proxyOptions.http || '');
        }
        this.invidiousClient = axios_1.default.create(config);
        // Store instance URLs for fallback
        this.invidiousClient.__instanceUrls = instanceUrls;
        this.invidiousClient.__currentInstanceIndex = 0;
    }
    /**
     * Configure proxy settings
     */
    setProxyOptions(options) {
        this.proxyOptions = { ...this.proxyOptions, ...options };
        this.initializeHttpClient();
        if (this.invidiousOptions.enabled) {
            this.initializeInvidiousClient();
        }
    }
    /**
     * Configure Invidious settings
     */
    setInvidiousOptions(options) {
        this.invidiousOptions = { ...this.invidiousOptions, ...options };
        if (this.invidiousOptions.enabled) {
            this.initializeInvidiousClient();
        }
        else {
            this.invidiousClient = null;
        }
    }
    /**
     * Fetch transcript with enhanced proxy support
     */
    async fetch(videoId, languages = ['en'], preserveFormatting = false, formatter) {
        // Use base API with enhanced proxy support
        return await this.baseApi.fetch(videoId, languages, preserveFormatting);
    }
    /**
     * List available transcripts
     */
    async list(videoId) {
        return await this.baseApi.list(videoId);
    }
    /**
     * Get video metadata
     */
    async getVideoMetadata(videoId) {
        // This would need to be implemented in the base API
        // For now, return a basic metadata object
        return {
            id: videoId,
            title: 'Video Title',
            description: 'Video Description',
            author: 'Video Author',
            channelId: 'channel_id',
            lengthSeconds: 0,
            viewCount: 0,
            isPrivate: false,
            isLiveContent: false
        };
    }
}
exports.EnhancedYouTubeTranscriptApi = EnhancedYouTubeTranscriptApi;
